File: src/domain/models/portfolio.py
"""Portfolio domain model — Position tracking + AI rescoring."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date


@dataclass
class Position:
    """A single holding in the portfolio."""
    stock_code: str
    stock_name: str
    shares: int              # 持仓数量
    cost_price: float        # 成本价
    current_price: float = 0.0
    market_value: float = 0.0       # 市值 = shares * current_price
    cost_value: float = 0.0         # 成本 = shares * cost_price
    profit_loss: float = 0.0        # 盈亏金额
    profit_loss_pct: float = 0.0    # 盈亏比例
    weight_pct: float = 0.0         # 占总仓比例
    ai_score: float = 50.0          # 当前 AI 评分
    ai_direction: str = "neutral"
    ai_signal: str = ""             # 最强信号名称
    risk_level: str = "中"
    added_date: str = ""            # 建仓日期
    last_score_change: float = 0.0  # 最近评分变化


@dataclass
class Portfolio:
    """Complete portfolio snapshot."""
    date: str = ""
    total_value: float = 0.0
    total_cost: float = 0.0
    total_pl: float = 0.0
    total_pl_pct: float = 0.0
    daily_pl: float = 0.0
    daily_pl_pct: float = 0.0
    cash: float = 0.0
    positions: list[Position] = field(default_factory=list)
    ai_summary: str = ""            # AI-generated portfolio summary
    risk_summary: str = ""          # Portfolio risk assessment
    top_performer: str = ""         # Best performing holding
    worst_performer: str = ""       # Worst performing holding

    @property
    def position_count(self) -> int:
        return len(self.positions)

    @property
    def avg_ai_score(self) -> float:
        if not self.positions:
            return 50.0
        return sum(p.ai_score * p.weight_pct for p in self.positions) / 100

    def to_dict(self) -> dict:
        return {
            "date": self.date,
            "total_value": round(self.total_value, 2),
            "total_cost": round(self.total_cost, 2),
            "total_pl": round(self.total_pl, 2),
            "total_pl_pct": round(self.total_pl_pct, 2),
            "daily_pl": round(self.daily_pl, 2),
            "daily_pl_pct": round(self.daily_pl_pct, 2),
            "cash": round(self.cash, 2),
            "position_count": self.position_count,
            "avg_ai_score": round(self.avg_ai_score, 1),
            "ai_summary": self.ai_summary,
            "risk_summary": self.risk_summary,
            "top_performer": self.top_performer,
            "worst_performer": self.worst_performer,
            "positions": [
                {
                    "stock_code": p.stock_code,
                    "stock_name": p.stock_name,
                    "shares": p.shares,
                    "cost_price": round(p.cost_price, 2),
                    "current_price": round(p.current_price, 2),
                    "market_value": round(p.market_value, 2),
                    "cost_value": round(p.cost_value, 2),
                    "profit_loss": round(p.profit_loss, 2),
                    "profit_loss_pct": round(p.profit_loss_pct, 2),
                    "weight_pct": round(p.weight_pct, 2),
                    "ai_score": round(p.ai_score, 1),
                    "ai_direction": p.ai_direction,
                    "ai_signal": p.ai_signal,
                    "risk_level": p.risk_level,
                    "added_date": p.added_date,
                    "last_score_change": round(p.last_score_change, 1),
                }
                for p in self.positions
            ],
        }

File: vscode-ext/src/pages/portfolio.ts
/** Portfolio Page v1.0 — Position tracking + AI rescoring. */

import { pageShell } from '../webview/layout';
import { BASE_URL } from '../constants';

export function buildPortfolioPage(data: any): string {
    const pf = data.portfolio || {};
    const positions = pf.positions || [];
    const plColor = (pf.total_pl || 0) >= 0 ? 'up' : 'down';
    const plSign = (pf.total_pl || 0) >= 0 ? '+' : '';

    const content = `
<!-- Portfolio Summary -->
<div class="grid4">
<div class="card"><h3>总资产</h3><div class="metric-value" style="font-size:24px">¥${((pf.total_value || 0) / 10000).toFixed(1)}万</div></div>
<div class="card"><h3>总盈亏</h3><div class="metric-value ${plColor}">${plSign}${(pf.total_pl || 0).toFixed(0)}</div><span class="text-sm ${plColor}">${plSign}${(pf.total_pl_pct || 0).toFixed(2)}%</span></div>
<div class="card"><h3>今日盈亏</h3><div class="metric-value ${(pf.daily_pl || 0) >= 0 ? 'up' : 'down'}" style="font-size:28px">${(pf.daily_pl || 0) >= 0 ? '+' : ''}${(pf.daily_pl || 0).toFixed(0)}</div><span class="text-sm">${pf.daily_pl_pct || 0 > 0 ? '+' : ''}${(pf.daily_pl_pct || 0).toFixed(2)}%</span></div>
<div class="card"><h3>AI评分</h3><div class="metric-value ${(pf.avg_ai_score || 50) >= 70 ? 'up' : (pf.avg_ai_score || 50) >= 50 ? 'warn' : 'down'}">${(pf.avg_ai_score || 50).toFixed(0)}</div><span class="text-sm text-muted">${pf.position_count || 0}只持仓</span></div>
</div>

<!-- AI Summary + Risk -->
<div class="grid2">
${pf.ai_summary ? `
<div class="card" style="border-left:3px solid #7C3AED">
<h3>AI 持仓分析</h3>
<p style="font-size:13px;line-height:1.6;color:#9CA3AF;margin-top:4px">${pf.ai_summary}</p>
</div>` : ''}
${pf.risk_summary ? `
<div class="card" style="border-left:3px solid #F59E0B">
<h3>风险评估</h3>
<p style="font-size:13px;color:#9CA3AF;margin-top:4px">${pf.risk_summary}</p>
${pf.top_performer ? `<div style="margin-top:8px;font-size:12px"><span style="color:#22C55E">最佳: ${pf.top_performer}</span></div>` : ''}
${pf.worst_performer ? `<div style="margin-top:4px;font-size:12px"><span style="color:#EF4444">最差: ${pf.worst_performer}</span></div>` : ''}
</div>` : ''}
</div>

<!-- Positions Table -->
<div style="padding:0 24px">
<div class="card">
<div class="card-header"><h3>持仓明细</h3><span class="text-sm text-muted">AI评分 · 每日更新</span></div>
<table>
<thead><tr>
<th>#</th><th>代码</th><th>名称</th><th>持仓</th><th>成本</th><th>现价</th><th>市值</th><th>盈亏</th><th>占比</th><th>AI评分</th><th>信号</th><th>风险</th>
</tr></thead>
<tbody>
${positions.map((p: any, i: number) => {
    const plClass = (p.profit_loss || 0) >= 0 ? 'up' : 'down';
    const plSign = (p.profit_loss || 0) >= 0 ? '+' : '';
    const scClass = (p.ai_score || 50) >= 70 ? 'up' : (p.ai_score || 50) >= 50 ? 'warn' : 'down';
    const riskColor = p.risk_level === '极低' || p.risk_level === '低' ? '#22C55E' :
                      p.risk_level === '中' ? '#F59E0B' : '#EF4444';
    const scChange = (p.last_score_change || 0);
    const changeArrow = scChange > 3 ? '↑' : scChange < -3 ? '↓' : '→';
    const changeColor = scChange > 0 ? '#22C55E' : scChange < 0 ? '#EF4444' : '#9CA3AF';

    return `<tr onclick="analyzeStock('${p.stock_code}')" style="cursor:pointer">
<td>${i + 1}</td>
<td class="stock-code">${p.stock_code}</td>
<td class="stock-name">${p.stock_name}</td>
<td>${p.shares}股</td>
<td>¥${(p.cost_price || 0).toFixed(2)}</td>
<td>¥${(p.current_price || 0).toFixed(2)}</td>
<td>¥${((p.market_value || 0) / 10000).toFixed(1)}万</td>
<td><span class="${plClass}" style="font-weight:600">${plSign}${(p.profit_loss || 0).toFixed(0)}</span><br><span class="text-sm ${plClass}">${plSign}${(p.profit_loss_pct || 0).toFixed(1)}%</span></td>
<td>${(p.weight_pct || 0).toFixed(1)}%</td>
<td><span class="${scClass}" style="font-weight:700;font-size:15px">${(p.ai_score || 50).toFixed(0)}</span> <span style="font-size:11px;color:${changeColor}">${changeArrow}</span></td>
<td><span class="tag tag-${p.ai_direction === 'buy' ? 'up' : p.ai_direction === 'sell' ? 'down' : 'info'}">${p.ai_signal || '-'}</span></td>
<td><span style="color:${riskColor};font-size:12px">${p.risk_level || '-'}</span></td>
</tr>`;
}).join('') || '<tr><td colspan="12" class="empty-state">暂无持仓数据</td></tr>'}
</tbody>
</table>
</div>
</div>

<div style="text-align:center;padding:16px" class="text-muted text-sm">
数据更新: ${pf.date || '--'} · AI评分每日自动刷新
</div>`;

    return pageShell('portfolio', 'Portfolio · 持仓中心', content, `
// Auto-refresh portfolio every 120s
setInterval(async () => {
    try { await fetch('${BASE_URL}/portfolio/overview'); } catch(e) {}
}, 120000);
`);
}
